File: practice/algorithms/sorting/the_full_counting_sort/the_full_counting_sort.py
#!/usr/bin/env python

def stable_sort(i_arr):
    count = [0 for _ in range(100)]
    for x in i_arr:
        count[x[0]] += 1
    
    total = 0
    for i in range(100):   # i = 0, 1, ... k-1
        oldCount = count[i]
        count[i] = total
        total += oldCount
    
    res = [0 for _ in range(len(i_arr))]
    for x in i_arr:
        res[count[x[0]]] = x
        count[x[0]] += 1
 
    return res

n = int(raw_input().strip())
phrases = []
for i in range(n):
    row = raw_input().strip().split()
    if i < n // 2:
        phrases.append([int(row[0]),"-"])
    else:
        phrases.append([int(row[0]),str(row[1])])

sorted_phrases = stable_sort(phrases)
for item in sorted_phrases:
    print item[1],

File: practice/algorithms/sorting/the_full_counting_sort/the_full_counting_sort.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <string>
#include <iostream>
#include <algorithm>

using namespace std;

vector<pair<int, string>> stable_sort(const vector<pair<int, string>> & i_arr)
{
    vector<int> count(100);
    
    vector<pair<int, string>>::const_iterator it = i_arr.begin();
    for (; it != i_arr.end(); ++it)
    {
        count[it->first]++;
    }
    
    int total = 0;
    for (int pos = 0; pos < 100; ++pos)
    {
        int c = count[pos];
        count[pos] = total;
        total += c;
    }
    
    vector<pair<int, string>> res(i_arr.size());
    
    it = i_arr.begin();
    for (; it != i_arr.end(); ++it)
    {
        res[count[it->first]] = *it;
        count[it->first]++;
    }
    
    return res;
}

int main()
{
    int n;
    cin >> n;
    
    vector<pair<int, string>> phrases(n);
    for (int pos = 0; pos < n; ++pos)
    {
        int x;
        string phrase;
        
        cin >> x >> phrase;
        
        if (pos < n / 2)
        {
            phrases[pos] = make_pair(x, "-");
        }
        else
        {
            phrases[pos] = make_pair(x, phrase);
        }
    }
    
    vector<pair<int, string>> sorted = stable_sort(phrases);
    
    vector<pair<int, string>>::const_iterator it = sorted.begin();
    for (; it != sorted.end(); ++it)
    {
        cout << it->second << " ";
    }
    
    return 0;
}

